hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/producer/kafka/KafkaBrokerTopicMetadataFetcher.java

```diff
@@ -0,0 +1,63 @@
+package pl.allegro.tech.hermes.frontend.producer.kafka;
+
+import org.apache.kafka.common.PartitionInfo;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import pl.allegro.tech.hermes.frontend.metric.CachedTopic;
+import pl.allegro.tech.hermes.frontend.producer.BrokerTopicMetadataFetcher;
+
+import java.util.List;
+
+public class KafkaBrokerTopicMetadataFetcher implements BrokerTopicMetadataFetcher {
+
+    private static final Logger logger = LoggerFactory.getLogger(KafkaBrokerTopicMetadataFetcher.class);
+
+    private final Producers producers;
+    private final KafkaTopicMetadataFetcher kafkaTopicMetadataFetcher;
+
+    public KafkaBrokerTopicMetadataFetcher(Producers producers, KafkaTopicMetadataFetcher kafkaTopicMetadataFetcher) {
+        this.producers = producers;
+        this.kafkaTopicMetadataFetcher = kafkaTopicMetadataFetcher;
+    }
+
+    @Override
+    public boolean tryFetchFromLocalDatacenter(CachedTopic cachedTopic) {
+        String kafkaTopicName = cachedTopic.getKafkaTopics().getPrimary().name().asString();
+
+        try {
+            List<PartitionInfo> partitionInfos = producers.get(cachedTopic.getTopic()).partitionsFor(kafkaTopicName);
+            if (anyPartitionWithoutLeader(partitionInfos)) {
+                logger.warn("Topic {} has partitions without a leader.", kafkaTopicName);
+                return false;
+            }
+            if (anyUnderReplicatedPartition(partitionInfos, kafkaTopicName)) {
+                logger.warn("Topic {} has under replicated partitions.", kafkaTopicName);
+                return false;
+            }
+            if (partitionInfos.size() > 0) {
+                return true;
+            }
+        } catch (Exception e) {
+            logger.warn("Could not read information about partitions for topic {}. {}", kafkaTopicName, e.getMessage());
+            return false;
+        }
+
+        logger.warn("No information about partitions for topic {}", kafkaTopicName);
+        return false;
+    }
+
+    @Override
+    public boolean tryFetchFromDatacenter(CachedTopic topic, String datacenter) {
+        // TODO: To fetch metadata for a selected datacenter https://github.com/allegro/hermes/pull/1823 is required.
```

Comment:
In the final implementation we have to be extra careful: this class uses `Producers` and `KafkaTopicMetadataFetcher`. `KafkaTopicMetadataFetcher` in turn uses `AdminClient`. So we are essentially using two different Kafka clients (KafkaProducer and AdminClient) here and we need to make sure in the future that both of them refer to the same datacenter

Reply:
Perhaps, an option worth considering is to introduce datacenter awareness in this PR since its scope is relatively narrow and it is easier to spot any bugs.

Reply:
Let me play around with this a bit. Perhaps remodeling this class slightly will address this comment and https://github.com/allegro/hermes/pull/1832#discussion_r1512650819.